File: sorting/quick_sort/quick_sort.go
package quick_sort

func quickSort(nums []int) {
	if len(nums) < 2 {
		return
	}

	pivot := partition(nums)
	quickSort(nums[:pivot])
	quickSort(nums[pivot+1:])
}

func partition(nums []int) int {
	pvt := len(nums) - 1 // the last idx as a pivot and sort through it
	pvtVal := nums[pvt]
	left := 0

	for right := 0; right < len(nums)-1; right++ {
		if nums[right] < pvtVal {
			nums[left], nums[right] = nums[right], nums[left]
		}
	}

	nums[left], nums[pvt] = nums[pvt], nums[left]
	return left
}

File: sorting/quick_sort/quick_sort.cpp
  #include <iostream>
  #include <vector>
  #include <algorithm>


  size_t findPivot(std::vector<int>& nums, size_t start, size_t end) {
    size_t pvt = end;
    int pvtVal = nums[pvt];

    size_t left = start;

    for (size_t right = start; right < pvt; ++right) {
      if (nums[right] < pvtVal) {
        std::swap(nums[right], nums[left]); 
        left++;
      }
    } 

    std::swap(nums[left], nums[pvt]);

    return left;
  }

  void quickSort(std::vector<int>& nums, size_t start, size_t end) {
    if (end - start < 1) {
      return;
    }

    size_t pivot = findPivot(nums, start, end);

    if (pivot > 0) {
      quickSort(nums, start, pivot-1);
    } 
    quickSort(nums, pivot+1, end);
  }
